[#13] 상품 수정, 삭제, 이미지

2eve2 opened the pull request:
상품 UD
상품 이미지 CRUD

2eve2 committed:
refactor: 이미지 api 분리

2eve2 committed:
refactor: api 주소 수정

- 명구님이 shop 구현 후, shopId 추가

2eve2 committed:
feat: 상품 삭제 api

2eve2 committed:
refactor: db 포트 포워딩, flyway ddl 네이밍

- 컨테이너의 TCP 포트 3306을 Docker 호스트 포트 13306에 매핑

2eve2 committed:
refactor: 이미지 컨트롤러 분리, 도메인 필드 추가

2eve2 committed:
feat: 로컬 이미지 업로더

2eve2 committed:
refactor: ALTER item_image table flyway V2.1

- multi pk, fk 설정
- 컬럼 삭제, 추가

2eve2 committed:
refactor: ALTER item_option table flyway V1.1

- fk 설정 추가
- 버전 관리 .X 형식 채택

2eve2 committed:
refactor: ItemImageRepository 불필요한 메서드 삭제

2eve2 committed:
refactor: delete multi table records

- foreign key constraint fails 문제 해결 필요

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/ItemImageController.java:
```diff
@@ -0,0 +1,29 @@
+package com.flab.livecommerce.presentation.item;
+
+import com.flab.livecommerce.application.item.facade.ItemImageManager;
+import com.flab.livecommerce.common.response.CommonApiResponse;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestPart;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.multipart.MultipartFile;
+
+@RequestMapping("/api/v1/item")
+@RestController
+public class ItemImageController {
+
+    private final ItemImageManager itemImageManager;
+
+    public ItemImageController(ItemImageManager itemImageManager) {
+        this.itemImageManager = itemImageManager;
+    }
+
+    @PostMapping("/{itemId}/image")
+    public CommonApiResponse uploadItemImage(
+        @RequestPart("thumbnailImg") MultipartFile thumbnailImage,
+        @RequestPart(value = "specificImg", required = false) MultipartFile[] specificImages
+    ) {
+        itemImageManager.uploadItemImage(thumbnailImage, specificImages);
+        return CommonApiResponse.success(null);
+    }
+}
```
ITEM 컨트롤러와의 RequestMapping 시작경로가 같은데 해당 부분 따로 컨트롤러 만드신 이유가 뭔지 궁금합니다.

2eve2 replied:
이미지 처리와 상품 정보 처리를 한 곳에서 관리하는 것보다 분리하는 것이 관리나 역할 측면상 더 좋다고 판단했습니다.
지금까지 저희가 설계한 비즈니스 로직상으로도 이미지와 상품 정보간에 의존성이 크게 필요하지 않기도 해서 분리하는 것이 더 좋을 것 같네요.

혹시 더 좋은 api 설계가 있어 의견 주시면 고려해보겠습니다~

dding94 replied:
그래도 ddd 개념으로 보면 이미지처리도 아이템이라는 집합에 속하는데 매니저와 컨트롤러까지 분리하실 필요가 있을까 생각이 들었습니다. 개인적인 의견은 Item 도메인은 itemManger 라는 하나의 통로를 통해서 구현을 하는게 오히려 더 관리하기 좋지 않을까.. 그렇기에 컨트롤러도 분리 될 필요가 없다고 생각을 해봅니다.

2eve2 replied:
cf. 밑에서 동일한 질문에 대해 답변하였습니다. 

상품 정보와 이미지에 대한 생성 및 변경이 다르며 api 및 aggregate도 분리해서 설계하였으므로 컨트롤러를 분리하였습니다.

dding94 commented on src/main/resources/db/migration/V1.1__alter_item_option_table_add_fk.sql:
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE item_option_group ADD FOREIGN KEY(item_id) REFERENCES item(id);
+
+ALTER TABLE item_option ADD FOREIGN KEY(item_option_group_id) REFERENCES item_option_group(id);
+
```
realMysql 99p 를 참고해보면 FK 제약 조건을 안걸고도 해결이 가능하지 않을까 라는 생각이 듭니다.
추가로 fk 설정시 인덱스가 필요한걸로 아는데 빼먹으신게 아닌지..
멘토님도 cascade 없이 구성해보면 좋다는 얘기를 하셔서 이게 조회나 delete 나 같은 맥락이지 않을까 생각이드네요

2eve2 replied:
fk 제약 조건을 걸지 않을 수도 있지만, 아이템 - 아이템 옵션 그룹 - 옵션처럼 아이템이 등록되고 조회되기 위해서 데이터 정합성이 반드시 지켜져야 하는 도메인의 경우에는 fk constraint를 정의하여 올바른 데이터를 유지하는 것이 좋다고 판단했습니다. 

innodb에서 외래 키 제약 조건이 있는 테이블을 생성하지만 지정된 열에 인덱스가 없는 경우 인덱스가 생성된다고 합니다.

그리고 fk 설정했다고 해서 cascade가 설정되는 것은 아니라고 알고 있는데, 조금 다른 맥락인 것 같습니다. 현재 cascade 설정은 하지 않고, 아이템 옵션 그룹, 옵션 모두 아이템의 id를 갖게 해서 삭제되게끔 구현 중 입니다.

https://dev.mysql.com/doc/refman/8.0/en/constraint-foreign-key.html

피드백 주셔서 찾아봤는데, 저랑 다르게 알고 계시네요. 혹시 제가 잘못된 정보를 공부한 것이라면 의견 주시면 감사하겠습니다:)

dding94 replied:
책에서는 데드락 발생 위험때문에 최대한 안거는게 좋다 얘기해서 안거는게 좋다고 생각을 했습니다.
저희 도메인에서는 그러면 말씀 주신대로 정합성을 위해 fk 설정을 해도 괜찮겠네요.

그리고 cascade 이 부분은 제가 잘 못 알고있었네요. 따로 설정을 하지 않는 이상은 적용이 안되네요.
제가 db를 더 많이 챙겨보긴 해야할 것 같습니다

2eve2 committed:
feat: item, itemOptionGroup 수정 api

- +) ItemOptionSeriesServiceImpl parameter 수정

2eve2 committed:
feat: itemOption 수정 api

dding94 commented on src/main/java/com/flab/livecommerce/application/item/command/RegisterItemOptionCommand.java:
```diff
@@ -16,9 +15,9 @@ public class RegisterItemOptionCommand {
     private Long price;
 
-    public ItemOption toEntity(ItemOptionGroup itemOptionGroup) {
+    public ItemOption toEntity(Long itemOptionGroupId) {
```
정정하겠습니다.
저장 로직에 도메인의 id만 사용하는것도 인지하고 있었고 나중에 Jpa 추가할거 고려해서 이렇게 구현했습니다.
혹시 이렇게 바꾸신 이유가 있을까요

2eve2 replied:
id만 필요한데 상품 수정 등의 작업 시 불필요한 entity가 파라미터로 전달이 되어 메서드 오버로딩을 해야할 것 같아 지웠습니다. 다른 조회, 삭제 등에서도 id만 쓰이기도 했고요.

JPA는 추후 적용될 수 있는 사항이라 해당 부분은 아직 공부가 덜 되었습니다. 

우선 jdbc template, mybatis를 적용 중이니 추후 jpa 적용할 때 해당 부분의 수정이 필요하면 바꿔도 될 것 같습니다.

dding94 replied:
메서드 오버로딩을 해야할 것 같다가 무슨 의미인지 잘 모르겠습니다. 좀더 자세한 설명 부탁드려도 될까요?

그부분을 떠나서 현재는 id를 받도록 되어있으니 id만 파라미터로 넘기도록 하는게 괜찮을 것 같습니다.

2eve2 replied:
toEntity() 메서드를 써야하는데, 받아와지는 파라미터 값이 달라져서 메서드 오버로딩 혹은 객체에서 일일이 id값을 가져와야 가능하다는 뜻이었습니다. 

근데 지금 delete table 설계가 달라지고 jpa도 고려해봤을 때 entity를 파라미터로 받도록 다시 수정하는게 어떤지 고려중입니다. 반영해보고 푸시해보겠습니다.

dding94 replied:
제가 잘 못 이해하고 있나 해서 여쭙는건데 Command 객체내에 toEntity 메서드는 하나인데 메서드 오버로딩이 이루어질 수 있나요?

2eve2 replied:
cf. 구두로 말씀드렸는데, 리뷰를 안달았네요. 

toEntity() 메서드에서 받는 파라미터가 달라져서 메서드 오버로딩을 해야한다는 뜻이었고, 관련 부분은 entity를 파라미터로 받도록 통일했습니다.

2eve2 committed:
feat: 아이템 이미지 전달방식 변경

2eve2 committed:
refactor: 아이템 이미지 업로드

- multi pk 수정
- local 업로드 후 db 저장 완료 (ordering 0)

2eve2 committed:
feat: 상품 삭제

- 상품 정보, 이미지 전체 삭제 by itemId

2eve2 committed:
refactor: 상품 옵션 그룹 수정

- 상품 수정 시 매핑 문제 해결 필요

2eve2 committed:
feat: 이미지 순서 추가

2eve2 committed:
feat: 이미지 삭제 in database

2eve2 committed:
fix: 이미지 삭제 jdbctemplate

2eve2 committed:
fix: 상세이미지 순서 변경

- jdbctemplate 추가 필요

2eve2 committed:
refactor: 옵션 그룹 업데이트

2eve2 committed:
refactor: 리뷰 반영

- contentUrl을 response body에 넣어서 전달

2eve2 committed:
refactor: resolve conflict

2eve2 committed:
style: checkstyle

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/request/ItemFormRequest.java:
```diff
@@ -14,7 +14,7 @@
 @AllArgsConstructor
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
-public class RegisterItemRequest {
+public class ItemFormRequest {
 
     private Long shopId;
```
해당 부분 회원가입과 수정이 구분되어 사용되지 않는 이름으로 사용되는데 요청이 다른부분이라 클래스도 분리를 시켜야 한다고 생각합니다.

dding94 commented on src/main/java/com/flab/livecommerce/application/item/command/ItemOptionFormCommand.java:
```diff
@@ -10,7 +10,7 @@
 @AllArgsConstructor
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
-public class RegisterItemOptionCommand {
+public class ItemOptionFormCommand {
```
여기도 동일합니다! (회원가입 /수정 요청 구분)

2eve2 committed:
chore: ddl 업데이트

2eve2 committed:
refactor: 상품 수정 리팩토링, 리뷰 반영

- request, command 별도 관리
- 코드 간결화 및 옵션 에러 해결

2eve2 committed:
refactor: price parameter

- fix 파라미터 null error
- Integer to Long

2eve2 committed:
refactor: 이미지 table

- composed pk
- 파일명에서 불필요한 값 제거

2eve2 committed:
refactor: 이미지 순서 request, query

2eve2 committed:
fix: 이미지 id find 에러

- ddl 수정 및 그에 따른 insert, update 쿼리 수정

2eve2 committed:
refactor: 이미지 삭제 api

- pk 변동에 따른 불필요한 파라미터 제거, delete repository 통합

dding94 commented on src/main/java/com/flab/livecommerce/application/item/UpdateItemProcessor.java:
```diff
@@ -0,0 +1,27 @@
+    @Transactional
+    public void execute(UpdateItemCommand command, Long id) {
+        Item updatedItem = itemRepository.update(command.toEntity(), id);
+        itemOptionSeriesService.update(command, updatedItem);
```
ddd 개념을 인용하면 아이템이 루트니까 아이템을 업데이트를 했을때 옵션들도 업데이트 하도록 만들면 합리적이지 않을까 생각을 해봅니다. main 브런치에 order 저장하는 부분 제가 그렇게 구현을 해봤는데 한번 참고해보시고 의견 남겨주시면 감사하겠습니다.
(현재 아이템 저장도 바꾸긴 해야하는데 jpa 적용하면서 할 예정이긴합니다!)

2eve2 replied:
DDD 개념 상 하나의 root에서 모두 처리하는 것이 더 좋을 것 같습니다. 말씀 주신 pr 참고해서 작업해보겠습니다.

+) 해당 부분은 JPA 적용하면서 함께 처리하는 것이 더 좋을 것 같아 JPA 적용 PR에서 작업하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/application/item/facade/ItemImageManager.java:
```diff
@@ -0,0 +1,42 @@
+        this.updateImagePriorityProcessor = updateImagePriorityProcessor;
+    }
+
+    public void upload(Long itemId, MultipartFile thumbnailImage, MultipartFile[] specificImages)
+        throws IOException {
+        uploadImageProcessor.execute(itemId, thumbnailImage, specificImages);
+    }
```
해당 부분 파라미터를 위에 ItemImageManager  생성자 파라미터 라인맞춘것 처럼 바꾸는게 저는 가독성 측면에서 좋다고 생각하는데 어떻게 생각하시나요?

2eve2 replied:
개인적으로 파라미터의 길이가 길지 않아서 가독성이 나쁘지 않다고 생각했었는데, 통일성 차원에서도 파라미터 라인을 맞추는 것이 좋을 것 같습니다. 반영하겠습니다~

dding94 commented on src/main/java/com/flab/livecommerce/application/item/facade/ItemImageManager.java:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.web.multipart.MultipartFile;
+
+@Service
+public class ItemImageManager {
```
다정님은 아이템이미지는 아이템과 다른 집합으로 분류되어 관리되어야 한다고 생각해서 이렇게 분리를 하신건지 궁금합니다. 아이템 이미지는 아이템과의 집합이라고 생각을 하는데 코드 구현 부분을 보니 저랑 살짝 다르게 생각을 하시는 것 같아서 여쭤봅니다..!
저는 아이템매니저는 아이템의 유스케이스를 관리하는 얘라고 생각을해서 하나의 매니저에 몰아두어 관리하는게 좋다고 생각을 합니다.
책임측면에서도 매니저는 창구의 역할을 제대로 한다고 생각을 하고 있고, 프로세서가 유스케이스의 책임을 가지고 있어 변경시에도 매니저를 바꾸는게 아니라 프로세서의 클래스 안의 내용만 바꾸면 되니까 더 합리적이지 않을까 생각을 해봅니다.

2eve2 replied:
저희가 상품 정보와 상품 이미지를 분리해서 정보를 받고 처리하도록 api를 설계했었습니다. 때문에 저는 상품과 상품 이미지의 애그리거트 경계를 나눠도 된다고 생각해서 작업한 상황입니다. 
상품과 이미지의 crud api가 모두 분리되어 있다면 반드시 같은 클래스로 작업되지 않아도 생각했습니다.

dding94 replied:
이 부분은 이미지에 대한 부분이 아이템과 변경이 다르고, api 도 분리해서 설계가 되어 다시 생각을 해보니 다정님 의견이 합리적인 것 같습니다. 따로 분리를 시키는게 더 합리적이겠네요!

dding94 commented on src/main/java/com/flab/livecommerce/application/item/UpdateImagePriorityProcessor.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+
+    public void execute(UpdateImageOrderCommand command) {
+        int requestOrderSize = command.getImageIdList().size();
+        for (int i = 0; i < requestOrderSize; i++) {
+            itemImageRepository.updateOrder(command.getImageIdList().get(i), command.getOrderingList().get(i));
+        }
+    }
+}
```
어플리케이션 레이어에는 수행할 작업을 정의하고 도메인 객체가 문제를 해결하게 해야하는 계층이라고 생각을 해서 로직같은 부분이 들어가는게 아닌 작업을 위임을 시키고 최대한 얇게 가져가야한다고 생각을 합니다. 그래서 for문 같은 부분을 제거하는 방향으로 다른 객체에게 위임을 하도록 코드를 만들면 어떨까요?

2eve2 replied:
네, 애플리케이션 레이어보다는 ordering 상태를 지니고 있는 도메인 객체가 상태를 바꾸는 책임을 갖는 것이 더 객체지향적일 것 같습니다. 반영하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/application/item/UploadImageProcessor.java:
```diff
@@ -1,43 +1,43 @@
+import org.springframework.transaction.annotation.Transactional;
 import org.springframework.web.multipart.MultipartFile;
 
 public class UploadImageProcessor {
```
해당 프로세서도 마찬가지로 어플리케이션에 레이어에서 상세한 구현 로직을 처리한다고 생각을 합니다.

2eve2 replied:
네 위와 같은 이유로 변경하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/application/item/UploadImageProcessor.java:
```diff
@@ -1,43 +1,43 @@
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+        if (thumbnailImage.isEmpty()) {
+            throw new ItemImageNotFoundException("썸네일 이미지는 필수입니다.");
```
제가 만들고 공유를 못한거 같아서 죄송합니다. 예외처리 부분도 common에서 baseException 을 선언하여 만들었습니다. 아이템 이미지 관련해서 이부분 상속하여서 만들도록 처리하면 괜찮지 않을까요?

2eve2 replied:
아 exception 형식을 바꾸셨군요. 반영하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/item/config/ItemImageProcessorConfig.java:
```diff
@@ -0,0 +1,35 @@
+    public UploadImageProcessor uploadImageProcessor(
+        ItemImageRepository itemImageRepository
+    ) {
+        return new UploadImageProcessor(itemImageRepository, new LocalUploader());
```
config 에 세군데에 new LocalUploader 가 사용되는데 하나로 묶는 방법은 어떨까요? 그러면 변경시에 세 라인을 수정하지 않고 한 부분만 수정을 하면 된다고 생각을 합니다. 그렇게 사용하는게 조금 더 합리적이지 않을까요?

f-lab-bright commented on docker/docker-compose.yml:
```diff
@@ -6,7 +6,7 @@ services:
     restart: on-failure
     ports:
-      - 3306:3306
+      - 13306:3306
```
요거 포트는 명구님이랑 두 분 합의해서 바꾸시는 거였죠?

2eve2 replied:
멘토링 때 포트포워딩 문제가 있어 공유드린 후 별도로 말씀 안드렸는데, 명구님께 다시 한번 공유 드리겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/item/DeleteItemProcessor.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Transactional
+    public void execute(Long id) {
+        itemImageRepository.deleteById(id);
```
DeleteItemProcessor인데, ItemImageRepository의 의존성을 가지고 있는게 불편할 수 있을 것 같은데, 그럴 때는 어떻게 해볼 수 있을까요??
혹은 item이 지워진 이후에 image는 꼭 동시에 지워지지 않아도 되기 때문에, db의 io를 줄이기 위해서 후처리로 item image를 삭제하기로 결정한다면 어떻게 처리해야 할까요?

2eve2 replied:
api와 애그리거트 경계를 분리해서 작업한 상황에서 DeleteItemProcessor만 item과 item image repository를 함께 의존하는 것이 신경쓰였었는데...
흠 그러면 item repository 하나만 의존받고, item과 image의 delete 메서드를 호출하는 새로운 repository 메서드를 만들면 될까요?
근데 이것도 결국 item repositoryAdaptor 안에서 image repository를 의존 받아야하고, item과 image가 동시에 지워지는 로직일텐데.. 고민입니다.

2eve2 replied:
스프링 이벤트 리스너를 사용하면 의존성을 분리해서 상품과 상품 이미지간에 결합도가 높은 문제를 해결할 수 있어 적용하였습니다. 나중에 db의 io를 줄이기 위해 상품이 삭제될 때 이미지가 동시에 삭제되지 않게끔 구현하더라도, 의존성이 분리되어있다면 변경 포인트를 줄일 수 있습니다. 현재는 @Async를 이용해서 간단하게 상품과 이미지 삭제 처리를 비동기로 구현하였는데, 추후 비동기와 eventual consistency에 대해 더 공부한 뒤 작업을 고도화해보도록 하겠습니다.

f-lab-bright commented on src/main/resources/db/migration/V2.1__alter_table_item_image.sql:
```diff
@@ -0,0 +1,7 @@
+ALTER TABLE item_image
+    MODIFY item_id bigint not null comment '상품 ID',
+    MODIFY url VARCHAR(2500) comment '이미지 주소',
+    MODIFY name VARCHAR(500) comment '이미지 파일 이름',
+    ADD COLUMN ordering tinyint(1) comment '정렬 순서';
+
+ALTER TABLE item_image ADD FOREIGN KEY(item_id) REFERENCES item(id);
```
한 피처에서 하나의 테이블에 대해서는 굳이 파일을 분리해서 버저닝을 할 필요는 없습니다.
피처 개발 중에 스키마가 변경되는 경우는 그냥 create table 문 자체를 고쳐서 올려주시는게 더 좋습니다.
로컬에서 개발하실 때는 flyway conflict를 해결하셔야되긴 하겠지만요 :)
flyway로 관리하는건 코드들이 병합된 이후에 버전을 잘 관리하기 위함이라는걸 기억하시면 좋을 것 같아요 :)
피처당 flyway 버전 파일 하나를 가능하면 지켜서 작업하시는게 바람직할 것 같습니다 :)

2eve2 replied:
아하.. 제가 flyway로 스키마 관리법을 잘못 이해한 것 같습니다. 병합되기 전에 같은 피처에서 스키마가 변경될 경우에는 굳이 버저닝을 할 필요가 없는거군요. 해당 부분 하나로 통합하겠습니다.

f-lab-bright commented on src/main/resources/db/migration/V1.1__alter_item_option_table_add_fk.sql:
```diff
@@ -0,0 +1,7 @@
+ALTER TABLE item_option_group ADD FOREIGN KEY(item_id) REFERENCES item(id);
+
+ALTER TABLE item_option
+    ADD COLUMN item_id bigint not null comment '상품 ID',
+    ADD FOREIGN KEY(item_option_group_id) REFERENCES item_option_group(id);
+
+ALTER TABLE item DROP COLUMN model_number;
```
이건 이 pr에서 했어야하는 작업은 아니었을 것 같긴 하네요 :)

2eve2 replied:
다른 PR을 보고 통일하려는 생각에 해당 PR에 맞지 않는 작업을 한 것 같습니다. 앞으로 PR의 작업을 고려하고 코드를 분리해서 작업하겠습니다.

f-lab-bright commented on src/main/resources/db/migration/V1.2__Alter_item_ddl_.sql:
```diff
@@ -1,2 +1 @@
 alter table item ADD COLUMN shop_id bigint not null comment '가게 ID';
-alter table item drop model_number;
```
그리고 이번 버전에 대한 파일 변경도 가능하면 하시지 않는게 좋습니다.
이러면 이 pr이 반영된 이후의 최신 main 브랜치 기준에서 실행되는 모든 개발자의 서버는 conflict가 날테니까요 :)

2eve2 replied:
아 명구님 ddl을 병합하다 conflict가 생겨서 제 ddl에서 겹치는 파트를 지웠었는데, 이전 버전이 아닌 제 ddl을 변경해서 추가 conflict를 내지 않는 것이 좋았겠습니다. 앞으로 주의하겠습니다!!

f-lab-bright commented on src/main/java/com/flab/livecommerce/presentation/item/response/SearchItemResponse.java:
```diff
@@ -12,17 +11,17 @@ public class SearchItemResponse {
     private String description;
-    private Integer price;
-    private Integer salesPrice;
+    private Long price;
+    private Long salesPrice;
```
가격이라 int에서 long으로 변경하신 것 같은데, 정말 이 지점에서 변경할 필요가 있었을까요??
int의 범위를 넘는 수치가 들어갈 일이 있을까요??

2eve2 replied:
가격이라 int보다는 long이 더 적합하다고 생각했고, 다른 코드 및 PR을 고려해 통일성을 맞추기 위해 작업했었습니다.
(코드 중 일부 객체의 price는 Long으로 되어있었고, order PR을 확인해보니 모두 Long으로 지정되어 있었습니다.)

통일성 이유 외에는 현재 시점에서 가격이 int의 -2147483648 ~ 2147483647 범위를 넘어갈 가능성이 적으므로 굳이 변경하지 않았어도 되었겠네요.. ㅎㅎ

f-lab-bright commented on src/main/java/com/flab/livecommerce/presentation/item/ItemImageController.java:
```diff
@@ -0,0 +1,53 @@
+    @PostMapping
+    public CommonApiResponse uploadItemImage(
+        @PathVariable("itemId") Long id,
+        @RequestPart("thumbnailImg") MultipartFile thumbnailImage,
+        @RequestPart(value = "specificImg", required = false) MultipartFile[] specificImages
```
thumbnailImg는 상품 목록에서 보일 작은 이미지고, specificImages는 상품 상세에 보여질 이미지의 목록인건가요??

2eve2 replied:
네 맞습니다! 
따라서 상품 대표 이미지로서 thumbnailImg는 상품 등록 시 한 개가 필수적으로 존재해야하며, specificImg는 여러 개가 저장되어 보여질 수 있도록 하였습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/item/image/LocalUploader.java:
```diff
@@ -0,0 +1,36 @@
+     * 임의의 로컬 경로
+     * // TODO 경로 변경
+     */
+    private static final String LOCAL_PATH = "C:/study/";
```
이건 서버의 설정으로 분리할 수 있게 만드셔야 할 것 같네요.
application.yml의 설정으로 기본 패스를 주입할 수 있게 구성해주세요 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/item/image/LocalUploader.java:
```diff
@@ -0,0 +1,36 @@
+            image.transferTo(new File(uploadPath));
+            return ItemImage.builder()
+                .itemId(itemId)
+                .url(uploadPath)
```
image의 url에 LOCAL_PATH가 전부 붙은 주소가 들어가도 괜찮을까요??
이 주소로 브라우저에서 이미지를 불러올 수 있을까요??
추후에 이미지에 대한 저장소가 변경된다면, 마이그레이션 하기 쉬운 설계일까요??
어떻게 하면 추후에 이미지를 저장하는 곳이 변경되는 경우에도 어려움 없이 전달할 수 있을까요??

2eve2 replied:
추후 이미지에 대한 저장소 위치가 변경되더라도 db를 다 변경하지 않게끔 db에는 base path를 제외한 이미지 경로만 저장되게끔 변경하였습니다.

f-lab-bright reviewed (commented):
로컬 디스크에 이미지를 저장하고 있는 상태이니, 이 파일을 읽어서 이미지를 브라우저에서 그려줄 수 있도록 하는 endpoint도 하나 슥삭 만들어보셔도 좋을 것 같네요 :)
그래야 itemImage의 url을 넣어두고 반환하는게 의미가 있을 것 같네요 :)

2eve2 committed:
refactor: 리뷰 반영

- 한 피처 안에서 변경되는 flyway 스키마 통합

2eve2 committed:
refactor: 리뷰 반영

- file location 서버 설정으로 분리
- 파라미터 코드스타일 적용
- config method 하나로 분리

2eve2 committed:
refactor: 리뷰 반영

- Spring Event Listener: 상품 삭제 시 상품 정보, 이미지 의존성 분리
- 상품 이미지 삭제 비동기 이벤트 처리 (추후 배치/메시지 큐 방식 등 고려)

2eve2 committed:
feat: JPA entity, 연관관계 매핑

- 같은 aggregate로 객체를 참조하는 경우 메서드, command, 등 수정
- 기본 다대일 관계, item-optionGroup-option은 양방향 연관관계 적용

2eve2 committed:
refactor: JPA - Jdbctemplate

- 하나의 aggregate에서 item save (~ing)
- 관련문제 수정 중

2eve2 committed:
Revert "refactor: JPA - Jdbctemplate"

This reverts commit 3ad5609c86c618db92767b6ef9bb00f8b2d724ec.

2eve2 committed:
Revert "feat: JPA entity, 연관관계 매핑"

This reverts commit 327826e573ede759c1234daf94c7e7e39add43ee.

2eve2 committed:
refactor: 리뷰 반영

- BaseException 형식 적용

2eve2 committed:
refactor: 리뷰 반영

- 객체지향코드 리팩토링 (image uplod application, image uploader)
- 객체 간 메시지 추가 (image, item)
- findById param 오류 수정
- 불필요한 코드 삭제, 체크 스타일

2eve2 committed:
refactor: 리뷰 반영

- db url에 basepath 제외한 url 저장 (마이그레이션 용이)

2eve2 committed:
feat: 업로드 이미지 endpoint 구현

2eve2 committed:
fix: 파일 업로드 path 설정 오류 해결

2eve2 committed:
fix: application.yml null error

2eve2 committed:
refactor: 업로드 이미지 endpoint

- 브라우저에서 이미지를 바로 확인할 수 있도록 구현

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/item/GetImageProcessor.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+
+    public ResponseEntity<Resource> execute(String uploadPath) throws IOException {
```
application layer가 ResponseEntity를 반환하는게 적절한가요??
이 의존성은 presentation의 의존성일 것 같네요.

2eve2 replied:
ImageUploader에서 모든 path 관리를 하도록 해서 httpheader와 resource 처리를 한꺼번에 해서 반환하려고 하는 바람에.. presentation layer의 역할과 ResponseEntity가 무엇인지 제대로 생각하지 않고 구현했습니다. 

presentation layer는 HTTP 요청을 application layer가 필요로 하는 형식으로 변환해서 application layer에 전달하고, application layer의 응답을 HTTP 응답으로 변환하여 전송합니다. HttpEntity 클래스를 상속받아 구현한 클래스가 RequestEntity, ResponseEntity 클래스이므로 presentation의 의존성으로 나타내야 하는게 맞습니다. 

layer architecture와 의존성 관련한 문제가 꽤 나오네요. ㅜ
Image Uploader 책임을 더 고민하고, layered architecture를 고려하고 공부해서 다시 수정해보겠습니다!

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/item/GetImageProcessor.java:
```diff
@@ -0,0 +1,37 @@
+        }
+
+        Path filePath = Paths.get(imageUploader.getFullPath(uploadPath));
+        HttpHeaders headers = new HttpHeaders();
+        headers.set("content-Type", Files.probeContentType(filePath));
```
http로 이미지를 조회할 수 있게 해주는 책임이 해당 위치에 있는게 적절할까요?

2eve2 replied:
위의 이유로 http로 이미지를 조회할 수 있게 하는 책임은 presentation layer에 있어야합니다. 수정하겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/presentation/item/ItemImageController.java:
```diff
@@ -0,0 +1,66 @@
+    /*
+     * 업로드 한 이미지 확인용 endpoint
+     */
+    @GetMapping("/{uploadPath}")
+    public ResponseEntity<Resource> getUploadItemImage(
+        @PathVariable String uploadPath
+    ) throws IOException {
+        return itemImageManager.get(uploadPath);
+    }
```
`/api/v1/item/{itemId}/image/image.jpeg` 같은 path가 될텐데, 이렇게 사용되는게 적절할까요??
image 조회를 위해서 어떤 item의 id인지를 알아야 할 필요가 있나요?

2eve2 replied:
image가 마이그레이션될 때, 일관된 uri를 유지하기 어려우므로 itemId가 포함되는 것이 적절하지 않다는 것을 배웠습니다. 현재는 필요없는 path를 삭제하였습니다.
api 설계 부분이 좀 어려운데 말씀해주신 대로 앞으로는 다른 서비스도 참고해보면서 설계해보겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/item/GetImageProcessor.java:
```diff
@@ -0,0 +1,37 @@
+
+public class GetImageProcessor {
+
+    private final ImageUploader imageUploader;
```
ImageUploader가 GetImageProcessor에서 사용되는게 조금 이상하지 않을까요??
이미지를 읽어오기 위한 작업에 Uploader라는 인터페이스가 의존되어야 할 필요가 있을까요??
isp에 위반되지 않을까요?

2eve2 replied:
ImageUploader라는 인터페이스 명칭보다는 FileStorageService 명칭이 더 알맞다고 생각해서 변경하였고, GetImage할 때 image가 실제로 저장된 path(ex. 로컬 c드라이브 포함 주소)가 필요해서 FileStorageService에서 저장된 path를 불러오도록 구현했습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/item/UploadImageProcessor.java:
```diff
@@ -1,43 +1,47 @@
+        ItemImage storedThumbnail = imageUploader.uploadImage(thumbnailImage);
+        storedThumbnail.addItem(item);
+        itemImageRepository.save(storedThumbnail);
 
-    // TODO 경로 변경 필요
-    private String getLocalPath() {
-        return "C:/study/img";
+        for (MultipartFile specificImage : specificImages) {
+            ItemImage storedSpecific = imageUploader.uploadImage(specificImage);
```
파일 저장 중에 실패하는 경우가 생기면 어떻게 될까요??

2eve2 replied:
파일 저장 중 생기는 에러는 FileStorageService 내에서 구현하였습니다!

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/item/image/LocalUploader.java:
```diff
@@ -0,0 +1,81 @@
+
+    private static String localBasePath;
+
+    @Value("${spring.servlet.multipart.location}")
+    private void setLocalBasePath(String temp) {
+        localBasePath = temp;
+    }
```
setter injection은 사용하지 않는게 좋습니다.
이 녀석도 constructor로 주입받고, 속성 자체도 외부에서 주입받을 수 있게 해야 테스트하기 복잡하지 않아집니다.
지금 이 클래스를 테스트하기 위해서 다른 경로를 지정해서 테스트해야 한다면 쉽게 할 수 있을까요??

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/item/image/LocalUploader.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Override
+    public List<URI> loadAll(Item item) {
```
ImageUploader가 가져야 할 행위에 적합한가요?

2eve2 replied:
ImageUploader를 구현할 때, 이미지를 업로드하는 것뿐만 아니라 file store하는데 필요한 전체적인 서비스를 처리하게끔 하려고 했었기에 load와 upload 등의 기능을 함께 모아두었습니다. (https://spring.io/guides/gs/uploading-files/  등의 자료 참고함)
![image](https://user-images.githubusercontent.com/58822617/201602148-7b29cd3a-36cf-4468-b121-cc12201cc1b1.png)

그런데 클래스와 인터페이스명을 ImageUploader로 지어서 적절하지 않은 행위로 보여지는 것 같은데, 명칭을 FileStorageService 등으로 변경하면 괜찮을까요?

f-lab-bright replied:
그건 다정님이 고민해보고 합리적이라 생각 되신다면 바꿔보시면 되죠 :)

2eve2 replied:
네, 명칭을 바꿔서 이미지를 업로드하는 것뿐만 아니라 전체적인 파일을 저장하고 관리하는 서비스로 보여지게 하는 것이 합리적일 것 같습니다..! 수정하겠습니다.

2eve2 replied:
수정)
현재는 image를 저장하는 서비스와 이미지 서버 uri를 리턴하는 부분을 분리해야 srp와 isp 등에 더 알맞은 설계인 것 같아 분리하였습니다.

f-lab-bright reviewed (commented):
item이 조회됐을 때 이미지를 읽어올 수 있는 주소가 적절히 반환되지 않을 것 같네요?
업로드나 수정 시점에만 실제로 로딩 가능한 주소가 반환되는건 의미가 없습니다.
필요한건, 실제로 주소가 반환되는 응답 모델에 로딩될 수 있는 이미지의 전체 uri가 반환될 수 있어야 합니다

2eve2 committed:
refactor: 리뷰 - imageuploader 명칭 변경

2eve2 committed:
refactor: 리뷰 - layer 의존성

2eve2 committed:
refactor: FileClient 인터페이스 분리 (틀만 구성, 수정중)

- 저장하는 것과 저장하는 path를 보여주는 것의 책임 분리
- full path 관련 문제는 아직 해결 못함

2eve2 committed:
refactor: 리뷰 - Uri Generator 분리

2eve2 committed:
refactor: item Image resultset 추가

2eve2 committed:
refactor: 리뷰 - image search

- image search 시 image url 리턴

2eve2 committed:
refactor: ImageProperties

2eve2 committed:
refactor: 코드 정리

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/request/ItemOrderRequest.java:
```diff
@@ -0,0 +1,17 @@
+    private List<Long> imageIdList;
+    private List<Integer> orderingList;
+
+    public UpdateImageOrderCommand toCommand() {
```
Request 이름이랑 toCommand 반환되는 이름이 매치가 잘 안돼서 흐름이 좀 헷갈립니다.
순서를 업데이트 하기위한 커맨드로 변경하여 사용하는 것 같은데  Request 객체 그대로 읽으면 아이템 주문 요청으로 해석되어
순서의 Order 보다는 주문에 가깝게 해석이 되네요.

2eve2 replied:
주문 도메인도 있어서 ItemOrderUpdate요청이 중의적으로 해석되는 것 같네요. 
UpdateImageOrderingRequest로 변경하겠습니다!

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/ItemImageController.java:
```diff
@@ -0,0 +1,77 @@
+        return CommonApiResponse.success(null);
+    }
+
+    @PutMapping("/api/v1/item/{itemId}/image/priority")
+    public CommonApiResponse updateImagePriority(
+        @RequestBody ItemOrderRequest orderingRequest
+    ) {
```
![스크린샷 2022-11-22 오후 2 32 39](https://user-images.githubusercontent.com/21376853/203231666-4046bd96-b357-4a47-8ace-c54ad6433850.png)
해당 부분 URI 에서 itemId 를 사용하지 않으면 제거해도 되지않을까 생각이 드는데 혹시 이렇게 체크 되는 부분 확인해보셨을까요??

2eve2 replied:
itemImageId를 Request로 받도록 Request에서 받는 데이터를 변경하였었는데 API 주소 설계에 수정을 못한걸 놓쳤네요! 감사합니다

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/image/local/LocalStorageService.java:
```diff
@@ -0,0 +1,63 @@
+@Slf4j
+public class LocalStorageService implements FileStorageService {
+
+    private final String basePath;
```
제가 어제 여기 이부분 잘못 말씀드린거 같습니다.
다시 생각해보니 애초에 해당 클래스가 LocalStroage 라는 이름이 붙어있기 때문에  path 를 주입 받는 형태가 아닌
해당 클래스가 직접 path. 를 가지고 있는 형태가 더 적합하다는 생각이 듭니다.

이렇게 사용하면 basePath 가 다른 저장소로 변경이 된다면, 해당 클래스의 이름은 Local 이지만 Local 을 가리키지 않을 수 있기 때문입니다.

2eve2 replied:
위의 코드 리뷰에서 얘기나왔듯이 base Path를 constructor로 주입받고, 속성 자체도 외부에서 주입받을 수 있게 해야 테스트하기 복잡하지 않아지기 때문에 주입받게끔 했습니다. 

base Path가 로컬이 아닌 다른 외부 저장소(ex. s3)가 될 경우 파일 업로드 방식이 바뀌기 때문에 FileStorageService를 implement하는 구현 클래스를 새로 만들어서 확장할 생각입니다.

dding94 replied:
제가 이부분 멘토링 복기 했는데 멘토링때 멘토님이 말씀하신 의도가 제가 말한걸로 해석했는데 다시 여쭤봐야겠네요.

2eve2 replied:
아아 그 부분은 basePath 자체를 저번에는 FileClient 클래스에서 메서드 호출해서 가져왔기 때문에, 메서드 호출이 아니라 변수를 직접 갖게끔 구현하라는 코멘트셨던걸로 기억합니다. File을 저장하는 basePath를 다른 곳에서 가져오는 것이 이상해서요

위에 조금 올려보시면 setter injection 쓰지 말라고 하시면서 basePath를 어떻게 할 것인가에 대해 리뷰 남겨주신 내용도 참고해보고 구현했습니다. 확인해보시고 더 좋은 방향 있으면 알려주세요.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/image/local/LocalStorageService.java:
```diff
@@ -0,0 +1,63 @@
+            throw new ItemImageNotFoundException();
+        }
+
+        String randomFileName = UUID.randomUUID().toString();
+        String originalFilename = image.getOriginalFilename();
+        String dbFilePath = createUploadFileName(originalFilename, randomFileName);
```
제가 잘 이해가 안돼서 그러는데 혹시 이 부분은 이미지를 업로드 처리를 하기위해 path 도 해당 부분에서 다 생성하도록 만드신걸까요? 제 의견은 책임을 분리 시킬 수 있겠다는 생각도 드는데 다정님은 어떻게 생각을 하시나요?

2eve2 replied:
File을 업로드(저장)하고 관리하는 역할과 책임을 FileStorageService에게 주었습니다.
그래서 File을 저장할 경로도 해당 서비스에서 만들고 관리하는 것이 합리적일 것이라 생각했습니다.
저장할 위치만 따로 만드는 책임을 분리하자는 말씀이신가요?

dding94 replied:
네 분리하면 괜찮지 않을까 하는데 만드실 때 의도를 그렇게 하셨으면 상관없을 것 같습니다.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/image/local/LocalUriGenerator.java:
```diff
@@ -0,0 +1,11 @@
+
+import com.flab.livecommerce.domain.image.FileUriGenerator;
+
+public class LocalUriGenerator implements FileUriGenerator {
+
+    @Override
+    public String getUriPrefix() {
+        return "http://localhost:8080/image/";
+    }
```
저는 해당 부분 Generator 라는 이름을 사용하면 메서드 이름은 generate 로 가는게 괜찮다고 생각이 들었습니다.
스프링프레임워크에서도 xxxGenerator 라는 인터페이스들은 generate 이름을 주로 사용해서 이게 더 문맥이 맞지 않을까요?
그리고 저희 프로젝트에서 TokenGenerator 도 generate 로 만들어서 통일시키는게 좋다고 생각이 듭니다.

2eve2 replied:
전체 Uri를 만드는 것이 아니라 uriPrifix만 가져오기 때문에 가져오는 목적을 명확히 보여주는 명칭을 썼었는데, 해당 클래스명을 UriPrefixGenerator로 바꾸고 generate 행동 명사 이름만 통일해서 사용하게끔 변경해도 좋을 것 같네요.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/image/persistence/ItemImageRepositoryAdapter.java:
```diff
@@ -0,0 +1,37 @@
+        this.itemImageRepository.updateOrdering(imageId, orderList);
+    }
+
+    @Override
+    public void updateOrder(Long imageId, Integer order) {
+        this.itemImageRepository.updateOrder(imageId, order);
+    }
```
해당 부분은 주문을 업데이트는 것 같은 명칭같아 보여서 다른 방향으로 수정하면 좋지않을까하는데 어떻게 생각하시나요?

2eve2 replied:
이미지를 주문하지는 않으나, 좀 헷갈릴 여지가 있을 수 있겠네요. 컬럼명대로 updateOrdering으로 바꾸겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/image/persistence/jdbctemplate/JdbcTemplateItemImageRepository.java:
```diff
@@ -0,0 +1,67 @@
+
+    public ItemImage save(ItemImage image) {
+        SqlParameterSource param = new BeanPropertySqlParameterSource(image);
+        Number key = jdbcInsert.executeAndReturnKey(param);
+        image.setId(key.longValue());
```
저는 해당 부분 long id = jdbcInsert.executeAndReturnKey(parameterSource).longValue(); 
로 바꿔서 image.setId(id) 이렇게 하는게 가독성이 더 좋다고 생각이 듭니다. 파라미터 명도 image 보다는 ItemImage 로 하는게 더 괜찮을 것 같기도 합니다. 
추가로 image, param 같은 줄임말 보다 parameter 를 사용하는게 저희가 줄임말을 쓰지 않기로 얘기를 나눠서 정했으니까 조금 더 코드가 길어지긴 하지만 parameter 로 사용하는게 좋지 않을까요?

2eve2 replied:
key를 가져오고, key의 id를 setId로 넣은거라 "id"가 아닌 "key"명칭의 number로 가져오면 저렇게 나눠서 코드를 작성하는 것이 명확하다고 생각했습니다. longValue와 함께 쓰려면 key가 아니라 id로 명칭을 바꾸고 Long으로 바꿔야겠네요. 

(추가) 이전에는 id값이 아닌 multi pk로 string, int 값을 둘다 가져와서 사용했으므로 key.~value()를 따로 작성했습니다.)

그리고 param이 parameter를 제가 임의로 줄인게 아니라 약어로 등록되고 사용되는 단어라서 사용한건데, 일단 변경하도록 하겠습니다!

dding94 replied:
key의 id 를 이라고 하니 잘 이해가 안되네요..  가져오는 Key 가 id 아닌가요??

2eve2 replied:
지금은 id로 key를 설정해놓게 바꿨는데 이전에는 itemImage table에서 id값 말고 다른 값들을 조합하여 multi pk를 두고 사용했어서 따로 분리해서 설명드렸습니다. 그래서 longValue()가 분리되었던거고요. key가 반드시 id값일 필요는 없으니 명칭도 바꿔야한다는 뜻입니다.

dding94 replied:
multi pk 를 두고 사용했으면 혹시 pk 가 string 으로 사용됐을까요?

2eve2 replied:
string이랑 int 형 두 개로 사용했었습니다. 지금은 id만 쓰므로 longValue()로만 가져오게끔 했어요.

dding94 replied:
그렇가면 executeAndReturnKey 의 반환 하는 key는 곧 id값 아닌가요? db에 접근해서 mysql이 생성한 key값을 Number 객체로 반환하는데 Number 객체는 string 이 될 수도 없고 또 개발자가 직접 키가 만든것이 될 수 없으니까 저는 id라고 하는게 맞는 것 같습니다.

2eve2 replied:
네 전에는 가독성을 고려해도 따로 타입을 두 개 지정해서 썼으므로 저렇게 구현한건데
현재는 multi pk가 아니라 id값을 쓰므로 longValue()쓰게끔 변경했고, 명칭과 타입 둘다 바꿔야겠다고 말씀드린거에요. 
병합한 코드 보시면 id, long으로 적용했습니다.

dding94 replied:
아 제가 이해를 잘못했군요! 죄송합니다

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/item/persistence/ItemOptionGroupRepositoryAdapter.java:
```diff
@@ -19,4 +19,25 @@ public ItemOptionGroupRepositoryAdapter(
+    public ItemOptionGroup update(ItemOptionGroup itemOptionGroup, Long id) {
+        return itemOptionGroupRepository.update(itemOptionGroup, id);
+    }
+
+
+
```
행 간격이 넓어 진 부분이 종종 보여서  구글체크스타일 적용한 단축키 한번씩 눌러서 확인하면 좋을 것 같아요.
저도 이부분 종종 많이 실수하는데 제것도 발견해주시면 고쳐주시면 감사하겠습니다.

2eve2 replied:
아 넵 현재 모듈화로 인한 컨플릭트때문에 pr ci 체크가 안되어서 놓쳤나봅니다. 감사합니다!

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/item/persistence/ItemOptionGroupRepositoryAdapter.java:
```diff
@@ -19,4 +19,25 @@ public ItemOptionGroupRepositoryAdapter(
+    public ItemOptionGroup update(Long itemId, List<ItemOptionGroup> itemOptionGroupList) {
+        return itemOptionGroupRepository.update(itemId, itemOptionGroupList);
+    }
+
```
필요 없는 주석이면 지우셔도 괜찮을 것 같아요!

dding94 reviewed (approved):
리뷰반영 할거 하시면 컨플릭트 해결 하고 병합하시면 될 것 같아요

2eve2 committed:
refactor: 코드 컨벤션 정리

2eve2 committed:
chore: 모듈화 시 이동할 파일 정리

2eve2 committed:
merge main
